File: flowersec-go/e2e/portable_protocol_vectors_test.go
package e2e_test

import (
	"context"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	controlplanehttp "github.com/floegence/flowersec/flowersec-go/controlplane/http"
	"github.com/floegence/flowersec/flowersec-go/fserrors"
	rpcv1 "github.com/floegence/flowersec/flowersec-go/gen/flowersec/rpc/v1"
	"github.com/floegence/flowersec/flowersec-go/observability"
	"github.com/floegence/flowersec/flowersec-go/proxy"
	"github.com/floegence/flowersec/flowersec-go/transportsecurity"
)

type portableProtocolVectors struct {
	Version                   int             `json:"version"`
	TransportPolicy           []transportCase `json:"transport_policy"`
	YamuxHeader               yamuxHeader     `json:"yamux_header"`
	RPCEnvelope               json.RawMessage `json:"rpc_envelope"`
	ControlplaneErrorEnvelope json.RawMessage `json:"controlplane_error_envelope"`
	ProxyHTTPRequestMeta      json.RawMessage `json:"proxy_http_request_meta"`
	ProxyWSOpenMeta           json.RawMessage `json:"proxy_ws_open_meta"`
	DiagnosticEvent           json.RawMessage `json:"diagnostic_event"`
}

type transportCase struct {
	URL            string   `json:"url"`
	Policy         string   `json:"policy"`
	AllowedHosts   []string `json:"allowed_hosts"`
	RiskAcceptance string   `json:"risk_acceptance"`
	Allowed        bool     `json:"allowed"`
}

type yamuxHeader struct {
	BytesHex string `json:"bytes_hex"`
	Version  byte   `json:"version"`
	Type     byte   `json:"type"`
	Flags    uint16 `json:"flags"`
	StreamID uint32 `json:"stream_id"`
	Length   uint32 `json:"length"`
}

type codeRegistry struct {
	Codes []struct {
		Code string `json:"code"`
	} `json:"codes"`
}

func TestPortableProtocolVectors(t *testing.T) {
	root := filepath.Join("..", "..")
	data, err := os.ReadFile(filepath.Join(root, "testdata", "portable_protocol_vectors.json"))
	if err != nil {
		t.Fatal(err)
	}
	var vectors portableProtocolVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	if vectors.Version != 1 {
		t.Fatalf("unexpected vector version: %d", vectors.Version)
	}

	for _, test := range vectors.TransportPolicy {
		var policy transportsecurity.Policy = transportsecurity.RequireTLS
		switch test.Policy {
		case "allow_plaintext_for_loopback":
			policy = transportsecurity.AllowPlaintextForLoopback
		case "allow_plaintext":
			policy = transportsecurity.AllowPlaintext
		case "network_plaintext":
			if test.RiskAcceptance != string(transportsecurity.PlaintextRiskAcceptPreE2ECredentialExposure) {
				t.Fatalf("invalid network plaintext risk acceptance %q", test.RiskAcceptance)
			}
			var err error
			policy, err = transportsecurity.NewNetworkPlaintextPolicy(transportsecurity.NetworkPlaintextPolicyOptions{
				AllowedHosts:   test.AllowedHosts,
				RiskAcceptance: transportsecurity.PlaintextRiskAcceptPreE2ECredentialExposure,
			})
			if err != nil {
				t.Fatalf("network plaintext policy: %v", err)
			}
		case "require_tls":
		default:
			t.Fatalf("unknown transport policy %q", test.Policy)
		}
		_, err := transportsecurity.Evaluate(
			context.Background(),
			test.URL,
			fserrors.PathDirect,
			transportsecurity.RuntimeNative,
			policy,
		)
		if (err == nil) != test.Allowed {
			t.Fatalf("transport case url=%q policy=%q allowed=%t: %v", test.URL, test.Policy, test.Allowed, err)
		}
	}

	header, err := hex.DecodeString(vectors.YamuxHeader.BytesHex)
	if err != nil {
		t.Fatal(err)
	}
	if len(header) != 12 ||
		header[0] != vectors.YamuxHeader.Version ||
		header[1] != vectors.YamuxHeader.Type ||
		binary.BigEndian.Uint16(header[2:4]) != vectors.YamuxHeader.Flags ||
		binary.BigEndian.Uint32(header[4:8]) != vectors.YamuxHeader.StreamID ||
		binary.BigEndian.Uint32(header[8:12]) != vectors.YamuxHeader.Length {
		t.Fatalf("unexpected Yamux header: %x", header)
	}

	var rpcEnvelope rpcv1.RpcEnvelope
	if err := json.Unmarshal(vectors.RPCEnvelope, &rpcEnvelope); err != nil {
		t.Fatal(err)
	}
	var rpcPayload struct {
		Message string `json:"message"`
	}
	if err := json.Unmarshal(rpcEnvelope.Payload, &rpcPayload); err != nil {
		t.Fatal(err)
	}
	if rpcEnvelope.TypeId != 7 || rpcEnvelope.RequestId != 42 || rpcPayload.Message != "flowersec" {
		t.Fatalf("unexpected RPC envelope: %+v", rpcEnvelope)
	}

	var controlplaneEnvelope controlplanehttp.ErrorEnvelope
	if err := json.Unmarshal(vectors.ControlplaneErrorEnvelope, &controlplaneEnvelope); err != nil {
		t.Fatal(err)
	}
	if controlplaneEnvelope.Error.Code != "artifact_not_found" {
		t.Fatalf("unexpected controlplane envelope: %+v", controlplaneEnvelope)
	}

	var httpMeta proxy.HTTPRequestMeta
	if err := json.Unmarshal(vectors.ProxyHTTPRequestMeta, &httpMeta); err != nil {
		t.Fatal(err)
	}
	if httpMeta.V != proxy.ProtocolVersion || httpMeta.Method != "POST" || httpMeta.TimeoutMS != 1500 {
		t.Fatalf("unexpected HTTP proxy metadata: %+v", httpMeta)
	}
	var wsMeta proxy.WSOpenMeta
	if err := json.Unmarshal(vectors.ProxyWSOpenMeta, &wsMeta); err != nil {
		t.Fatal(err)
	}
	if wsMeta.V != proxy.ProtocolVersion || wsMeta.ConnID != "connection-vector-1" {
		t.Fatalf("unexpected WebSocket proxy metadata: %+v", wsMeta)
	}

	var diagnostic observability.DiagnosticEvent
	if err := json.Unmarshal(vectors.DiagnosticEvent, &diagnostic); err != nil {
		t.Fatal(err)
	}
	if diagnostic.Path != "tunnel" || diagnostic.Stage != observability.DiagnosticStageYamux || diagnostic.Code != "liveness_timeout" {
		t.Fatalf("unexpected diagnostic event: %+v", diagnostic)
	}
	assertRegistryContains(t, filepath.Join(root, "stability", "connect_error_code_registry.json"), "resource_exhausted")
	assertRegistryContains(t, filepath.Join(root, "stability", "connect_diagnostics_code_registry.json"), diagnostic.Code)
}

func assertRegistryContains(t *testing.T, path string, code string) {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var registry codeRegistry
	if err := json.Unmarshal(data, &registry); err != nil {
		t.Fatal(err)
	}
	for _, entry := range registry.Codes {
		if entry.Code == code {
			return
		}
	}
	t.Fatalf("registry %s does not contain %q", path, code)
}

File: flowersec-ts/src/portable_protocol_vectors.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, test } from "vitest";

import { assertRpcEnvelope } from "./gen/flowersec/rpc/v1.gen.js";
import type { DiagnosticEvent } from "./observability/observer.js";
import type { HttpRequestMetaV1, WsOpenMetaV1 } from "./proxy/types.js";
import {
  AllowPlaintext,
  AllowPlaintextForLoopback,
  createNetworkPlaintextPolicy,
  PlaintextRiskAcceptance,
  enforceTransportSecurity,
  RequireTLS,
  type TransportSecurityPolicy,
} from "./client-connect/transportSecurity.js";
import type { ControlplaneErrorEnvelope } from "./controlplane/request.js";
import { decodeHeader, encodeHeader } from "./yamux/header.js";

type PortableVectors = Readonly<{
  version: number;
  transport_policy: ReadonlyArray<Readonly<{
    url: string;
    policy: "require_tls" | "allow_plaintext_for_loopback" | "allow_plaintext" | "network_plaintext";
    allowed_hosts?: readonly string[];
    risk_acceptance?: string;
    allowed: boolean;
  }>>;
  yamux_header: Readonly<{
    bytes_hex: string;
    version: number;
    type: number;
    flags: number;
    stream_id: number;
    length: number;
  }>;
  rpc_envelope: unknown;
  controlplane_error_envelope: ControlplaneErrorEnvelope;
  proxy_http_request_meta: HttpRequestMetaV1;
  proxy_ws_open_meta: WsOpenMetaV1;
  diagnostic_event: DiagnosticEvent;
}>;

type CodeRegistry = Readonly<{
  codes: ReadonlyArray<Readonly<{ code: string }>>;
}>;

const vectors = readJSON<PortableVectors>("../../testdata/portable_protocol_vectors.json");

describe("portable protocol vectors", () => {
  test("match transport, Yamux, RPC, controlplane, proxy, and diagnostic contracts", async () => {
    expect(vectors.version).toBe(1);

    for (const item of vectors.transport_policy) {
      const operation = enforceTransportSecurity({
        rawUrl: item.url,
        path: "direct",
        policy: transportPolicy(item),
      });
      if (item.allowed) {
        await expect(operation).resolves.toBeUndefined();
      } else {
        await expect(operation).rejects.toMatchObject({
          path: "direct",
          stage: "validate",
          code: "transport_policy_denied",
        });
      }
    }

    const headerBytes = decodeHex(vectors.yamux_header.bytes_hex);
    const header = decodeHeader(headerBytes, 0);
    expect(header).toEqual({
      version: vectors.yamux_header.version,
      type: vectors.yamux_header.type,
      flags: vectors.yamux_header.flags,
      streamId: vectors.yamux_header.stream_id,
      length: vectors.yamux_header.length,
    });
    expect(Array.from(encodeHeader(header))).toEqual(Array.from(headerBytes));

    const rpc = assertRpcEnvelope(vectors.rpc_envelope);
    expect(rpc).toMatchObject({
      type_id: 7,
      request_id: 42,
      response_to: 0,
      payload: { message: "flowersec" },
    });
    expect(vectors.controlplane_error_envelope.error).toEqual({
      code: "artifact_not_found",
      message: "No artifact is available.",
    });
    expect(vectors.proxy_http_request_meta).toMatchObject({
      v: 1,
      method: "POST",
      timeout_ms: 1500,
    });
    expect(vectors.proxy_ws_open_meta).toMatchObject({
      v: 1,
      conn_id: "connection-vector-1",
    });
    expect(vectors.diagnostic_event).toMatchObject({
      path: "tunnel",
      stage: "yamux",
      code_domain: "event",
      code: "liveness_timeout",
      result: "fail",
    });

    expect(registryCodes("../../stability/connect_error_code_registry.json")).toContain("resource_exhausted");
    expect(registryCodes("../../stability/connect_diagnostics_code_registry.json")).toContain(vectors.diagnostic_event.code);
  });
});

function transportPolicy(item: PortableVectors["transport_policy"][number]): TransportSecurityPolicy {
  switch (item.policy) {
    case "require_tls": return RequireTLS;
    case "allow_plaintext_for_loopback": return AllowPlaintextForLoopback;
    case "allow_plaintext": return AllowPlaintext;
    case "network_plaintext":
      if (item.risk_acceptance !== PlaintextRiskAcceptance.acceptPreE2ECredentialExposure) {
        throw new Error("invalid network plaintext risk acceptance");
      }
      return createNetworkPlaintextPolicy({
        allowedHosts: item.allowed_hosts ?? [],
        riskAcceptance: PlaintextRiskAcceptance.acceptPreE2ECredentialExposure,
      });
  }
}

function decodeHex(value: string): Uint8Array {
  if (value.length % 2 !== 0) throw new Error("hex input must have an even length");
  const output = new Uint8Array(value.length / 2);
  for (let i = 0; i < output.length; i++) {
    const byte = Number.parseInt(value.slice(i * 2, i * 2 + 2), 16);
    if (!Number.isFinite(byte)) throw new Error("invalid hex input");
    output[i] = byte;
  }
  return output;
}

function registryCodes(path: string): string[] {
  return readJSON<CodeRegistry>(path).codes.map((entry) => entry.code);
}

function readJSON<T>(path: string): T {
  return JSON.parse(readFileSync(new URL(path, import.meta.url), "utf8")) as T;
}
